_posts/2020-07-12-ray-tune.md

```diff
@@ -0,0 +1,457 @@
+initialize models with random selection. So, I will measure the time taken to
+complete `search.fit(X, y)` for various searches. That means it the data and
+model don't matter. Let's create a dummy model that takes 1 second for a
+`partial_fit` call with 50,000 examples:
+
+``` python
+from scipy.stats import uniform
+from sklearn.model_selection import make_classification
+from benchmark import ConstantFunction  # custom module
+
+model = ConstantFunction(latency=1 / 50e3, max_iter=max_iter)
+params = {"value": uniform(0, 1)}
+X_train, y_train = ...  # dummy data with 60,000 examples
+```
+
+This model will take 2 minutes to train for 100 epochs (aka passes through the
+data).  Details can be found at [stsievert/dask-hyperband-comparison][sts-dhc].
+Let's configure our searches to use 8 workers:
+
+``` python
+from sklearn.model_selection import RandomizedSearchCV, ShuffleSplit
+split = ShuffleSplit(test_size=0.2, n_splits=1)
+kwargs = dict(cv=split, refit=False)
+
+search = RandomizedSearchCV(model, params, n_jobs=8, n_iter=n_params, **kwargs)
+search.fit(X_train, y_train)  # 10.44 minutes
+```
+
+This search takes 10.44 minutes, a fairly long time to train with a model that
+takes 2 minutes to train. Let's try to beat that with Dask-ML and Ray's
+tune-sklearn. As mentioned [above](#consistency-with-the-scikit-learn-api),
+both Dask-ML and Ray have Scikit-learn compatible model selection API:
+
+``` python
+from dask_ml.model_selection import HyperbandSearchCV
+dask_search = HyperbandSearchCV(
+    model, params, test_size=0.2, max_iter=max_iter, aggressiveness=4
+)
+
+from tune_sklearn import TuneSearchCV
+ray_search = TuneSearchCV(
+    model, params, n_iter=n_params, max_iters=max_iter, early_stopping=True, **kwargs
+)
+
+dask_search.fit(X_train, y_train)  # 1.79 minutes
+ray_search.fit(X_train, y_train)  # 1.55 minutes
+```
+
+<img src="/images/2020-model-selection/n_workers=8.png" width="450px"
+/>
+
+[sts-dhc]:https://github.com/stsievert/dask-hyperband-comparison
+
+Clearly, Ray and Dask-ML offer similar performance for 8 workers when compared
+with Scikit-learn. Even relatively, Dask-ML is only 15% slower than Ray with 8
+workers. Ray might perform better with more workers because they implement an
+asychronous variant of Dask-ML's Hyperband; however, Dask-ML can scale up to 24
+workers for a search in "[Better and faster hyperparameter optimization with
+Dask][db-bf]." That's the number of GPUs on 4 nodes on the [Summit
+supercomputer][summit]!
+
+[summit]:https://en.wikipedia.org/wiki/Summit_(supercomputer)#Design
+
+The benchmark above ignores the fact that the Dask-ML implementation
+prioritizes fitting high scoring models, which is especially relevant in the
+serial context.[^priority] If there are 25 models to fit and only 5 slots to
+run them in, Dask-ML selects the models that have the highest score. See
+"[Better and faster hyperparameter optimization with Dask][db-bf]" for
+benchmarks.
```

Comment:
Might be worth linking to Matt's benchmarking post, as a general comment on the difficulty of these kinds of shootouts.